[1단계 - 출석 구현] 후유(김대동) 미션 제출합니다.

eoehd1ek opened the pull request:
## 체크 리스트
- [X] 미션의 필수 요구사항을 모두 구현했나요?
- [X] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [X] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr/)에 셀프 체크를 작성했나요?
- https://prolog.techcourse.co.kr/studylogs/3946

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.
-  1 (전혀 충족하지 못함)
-  2
- [X] 3 (보통)
-  4
-  5 (완벽하게 충족)
- 
### 선택한 점수의 이유를 적어주세요.
Indent depth 1을 지키지 못한 코드가 있었습니다.
Stream을 사용할 때 메서드마다 개행을 해 주는 컨벤션을 완벽하게 지켰는지 확인하지 못했습니다.
커밋 전 수정한 파일들에 대해 코드 정렬을 사용하지 않았습니다.


## 질문!
한 예시로 컨트롤러가 자식 도메인을 생성해서 넘겨줘야 하는 방식이 좋을지
아니면 컨트롤러가 자식 도메인을 생성하는데 필요한 데이터를 넘겨줘서,
부모 도메인이 자식 도메인을 생성하는 책임도 가지도록 해야할지 궁금합니다.

예) AttandenceBook에 있는 Some() 메서드가 Some(String name, String time) 처럼
Attend객체를 생성하기 위한 값인 String name, String time을 전달받고, Some 메서드 내에서 Attend를 생성하는 게 좋을지,
Some() 메서드가 Some(Attend attend) 처럼 Attend를 생성하고 넘겨주는 것이 좋을지 궁금합니다.
아니면 위 상황을 해결하기 위한 방법이나 패턴이 있을까요??
테스트 코드를 작성할 때 편했던 건 뒤에 말한 객체를 생성해서 전달하는 편이 쉬웠습니다.

## 어떤 부분에 집중하여 리뷰해야 할까요?
사실 제가 무엇을 놓치는지, 무엇을 모르는지 잘 모르겠습니다.
특정 코드를 어떠한 관점에서 보고 생각해 보면 좋을지 지적 해 주시면 열심히 찾아 공부하겠습니다.

jhan0121 committed:
docs: 요구사항 및 기능사항 추가

jhan0121 committed:
feat(Attend): 출석 관련 데이터를 관리하는 객체 추가

jhan0121 committed:
feat: 출석 가능 일자 확인 기능 추가

jhan0121 committed:
feat: 출석 추가 기능 구현

jhan0121 committed:
docs: CSV 고려 사항 추가

jhan0121 committed:
feat: 쉬는날 출석 시 예외 구현 추가

jhan0121 committed:
feat: 같은 날짜에 출석이 존재하는지 확인하는 기능 추가

jhan0121 committed:
feat: 출석 수정 기능 추가

jhan0121 committed:
feat: 대상 닉네임의 출석 수정 기능 추가

jhan0121 committed:
feat: 운영 시간 외 시간을 입력했을 경우 예외 처리 추가

jhan0121 committed:
fix: attend를 수정하면서 실패한 테스트 수정

jhan0121 committed:
test: findByDay 누락 테스트 추가

jhan0121 committed:
feat: 현재 날짜 이전까지 출석을 조회하는 기능 추가

jhan0121 committed:
feat: 닉네임 대상의 어제까지의 출석 정보를 반환하는 기능 추가

jhan0121 committed:
docs: csv 타입의 출석 데이터 파일 추가

jhan0121 committed:
feat: csv 타입의 파일을 읽어서 AttendanceBook 을 반환하는 기능

jhan0121 committed:
feat: Attend 와 LateTime, AbsenceTime 을 기반으로 출결을 판정하는 기능

jhan0121 committed:
feat: 이름과 날짜를 기반으로 출석과 출결 현황을 반환하는 기능 추가

jhan0121 committed:
feat: 출석, 지각, 결석 횟수를 바탕으로 제적 위험자를 파악하는 기능 추가

jhan0121 committed:
fix: 멤버 변수 타입(int -> long) 수정

jhan0121 committed:
feat: 출석 기록에서 출석, 지각, 결석 횟수를 파악하는 기능

jhan0121 committed:
refector: AttendanceResults를 일급 컬렉션으로 사용하도록 수정

jhan0121 committed:
feat: 출석부에서 제적 위험자 조회 기능 추가

jhan0121 committed:
docs: 누락한 기능 구현 사항 수정

jhan0121 committed:
feat: 출석부에 등록되지 않은 크루 닉네임을 사용하면 예외 발생 기능 추가

jhan0121 committed:
refactor: 역할에 맞는 패키지 이동

jhan0121 committed:
feat: InputView 추가

jhan0121 committed:
feat: OutputView 추가

jhan0121 committed:
feat: Controller 구현 및 기능 통합

summerlunaa commented on README.md:
```diff
@@ -1,3 +1,88 @@
 
 출석 미션 저장소
+
+## 요구 사항
```
꼼꼼한 요구사항 정리 좋습니다 👍

summerlunaa commented on src/main/java/view/OutputView.java:
```diff
@@ -0,0 +1,124 @@
+        if (warningStatus == WarningStatus.EXPEL) {
+            return "제적 대상자입니다.";
+        }
+        return "";
```
`formatWarningStatusShort` 메서드를 활용해볼 수도 있겠네요~

eoehd1ek replied:
리뷰 참고하여 formatWarningStatusShort 메소드를 사용하도록 수정했습니다.
알려주셔서 감사합니다.

[refector: 경고 대상자 출력시 중복 사용 코드를 삭제하도록 수정](https://github.com/woowacourse/java-attendance/pull/56/commits/f5fa77602da630515ef5f8aa3d1d255b2922329d)

summerlunaa commented on src/main/java/util/FileUtil.java:
```diff
@@ -0,0 +1,24 @@
+
+public class FileUtil {
+
+    private FileUtil() {
+
+    }
+
+    public static List<String> readFile(URL fileURL) {
+        try (BufferedReader reader = Files.newBufferedReader(Paths.get(fileURL.toURI()))) {
+            return reader.lines().skip(1).toList(); // column name이 존재하는 경우, 제거하기 위함
+        } catch (IOException | URISyntaxException e) {
+            throw new IllegalStateException(String.format("경로 문제: %s", fileURL.getPath()));
+        }
+    }
```
- private 생성자로 불필요한 인스턴스 생성을 잘 막았네요 👍 
- try-with-resources 사용 👍

summerlunaa commented on src/main/java/domain/WarningStatus.java:
```diff
@@ -0,0 +1,5 @@
+
+public enum WarningStatus {
+    CLEAR, WARNING, INTERVIEW, EXPEL
+}
```
WarningStatus, AttendStatus 등을 Enum으로 잘 관리하고 있네요 👍

summerlunaa commented on src/main/java/domain/Attends.java:
```diff
@@ -0,0 +1,79 @@
+                .filter(attend -> attend.isDayEqual(day))
+                .findFirst()
+                .orElse(Attend.fromDay(day));
+        // TODO: 예외를 던지던 테스트 코드 수정
```
어떤 주석일까요~?

eoehd1ek replied:
리팩터링 이전에는 해당하는 결과가 없는 경우 .orThrow 스트림을 사용해 예외를 던지는 동작을 가졌는데요.
리팩터링 이후에는 항상 Attend를 반환하는 동작으로 변경되었습니다.
이에 테스트 코드에 예외가 발생하는지 확인 테스트 코드를 지우게 되었습니다.
테스트 코드를 삭제하고 TODO 주석을 지우지 않았네요. 😭 

리팩터링 이후 항상 Attend를 반환하는 동작에 대한 테스트도 누락하여 주석 삭제와 함께 함께 커밋 남겼습니다.
[test: 날짜가 일치하는 출석 객체를 찾지 못했을 때 동작하는 누락된 테스트 추가](https://github.com/woowacourse/java-attendance/pull/56/commits/a5abdd5ad4a348b887b0bcc20c25536a4f38b8d2)
놓친 부분 알려 주셔서 감사합니다.

summerlunaa commented on src/main/java/domain/Attends.java:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    public void edit(Attend attend) {
+        if (hasDayEqualsAttend(attend)) {
+            // 똑같은 거 찾기
+            Attend before = findByAttend(attend);
+
+            // 찾은 객체를 토대로 list에서 삭제
+            removeByAttend(before);
+        }
+        addAttend(attend);
```
값이 존재하든 안 하든 remove 호출하고 add하면 결과는 똑같지 않나요? (없으면 remove 호출해도 아무 일도 안 일어나니..)
```suggestion
        removeByAttend(before);
        addAttend(attend);
```

eoehd1ek replied:
그 당시 removeByAttend 메서드를 작성하고 그 다음 edit 메서드를 작성 했을텐데요... 
정말 방금 작성한 코드인데 removeByAttend가 날짜와 시간이 동일한 Attend 객체를 필터링 한다고 생각했습니다.
그래서 기존 Attend 객체를 찾아오는 코드를 작성했습니다. 🙀

리팩터링으로 불필요한 로직인 Attend를 찾아오는 과정을 지우고, removeByAttend 메서드를 동작을 더 잘 설명할 수 있는 removeByDayEqualAttend 이름으로 변경했습니다.
[refector(Attends): 로직 상 필요 없는 코드를 제거하도록 수정](https://github.com/woowacourse/java-attendance/pull/56/commits/8141de7867afbfa2bff3868e609803e2e86fe8db)
더 나은 코드를 위한 제안 감사합니다.

summerlunaa replied:
`removeByDayEqualAttend`로 변경된 메서드명을 보니 차라리 이름을 `removeByDay`로 변경하고 Attend 대신 `int day`를 받는 것도 방법일 것 같네요ㅎㅎ

summerlunaa commented on src/main/java/domain/AttendCount.java:
```diff
@@ -0,0 +1,25 @@
+public record AttendCount(long attend, long late, long absence) {
+    public WarningStatus judgeWarning() {
+        long totalAbsenceCount = calculateTotalAbsenceCount();
+        if (totalAbsenceCount > 5) {
+            return WarningStatus.EXPEL;
+        }
+        if (totalAbsenceCount >= 3) {
+            return WarningStatus.INTERVIEW;
+        }
+        if (totalAbsenceCount >= 2) {
+            return WarningStatus.WARNING;
+        }
+        return WarningStatus.CLEAR;
```
totalAbsenceCount 기준은 WarningStatus에서 상수로 관리하지 않는 이유가 따로 있을까요?

eoehd1ek replied:
마감 시간이 다가와 먼저 구현을 완료하고자 매직 넘버 사용하였고, 책임을 분리하지 않았습니다.

리팩토링으로 totalAbsenceCount 기준의 책임을 WarningStatus에게 넘기고,
WarningStatus를 구하는 책임을 AttendCount에서 WarningStatus로 이동했습니다.
(AttendCount는 출석, 지각, 결석 횟수의 정보만을 다루는 책임이 있다고 생각했습니다.)
(WarningStatus 경고를 계산에 관련된 것을 처리하는 책임이 있다고 생각해, 경고 횟수를 가지게 했고, 위 AttendCount에서 제공하는 횟수의 정보를 토대로 어떤 경고 상태인지 판단하는 로직을 충분히 가질 수있다고 생각했습니다.)
[refector: 제적 위험 구하기 로직을 WarningStatus 클래스가 가지도록 수정](https://github.com/woowacourse/java-attendance/pull/56/commits/0dcd62be86028c9290d7f114642da84400285dd0)
좋은 질문 감사합니다.

summerlunaa replied:
<img width="529" alt="image" src="https://github.com/user-attachments/assets/a7dad0dc-5bf0-4ef4-95e6-67c1ebf1748d" />

너무 좋습니다 👍 위 노란줄만 확인해주세요.

summerlunaa commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,106 @@
+        this.outputView = outputView;
+    }
+
+    public void run() {
```
컨트롤러는 메서드 분리가 필요하겠네요! command도 상수로 관리할 수 있을 것 같아요 :)

eoehd1ek replied:
상수 관리 방법 의견 감사합니다.
컨트롤러 로직에 메서드 분리를 적용했습니다.
[refector: 커맨드 처리를 메서드로 분리하도록 수정](https://github.com/woowacourse/java-attendance/pull/56/commits/bb6723c154a46f8a3e5e87492c43285cf85f90fb)

```java
private boolean executeCommandAndCheckLoopContinue(String command, AttendanceBook attendanceBook) {
        if (command.equals(QUITE_COMMAND)) {
            return false;
        }
        if (command.equals(ATTEND_COMMAND)) {
            executeAttendCommand(attendanceBook);
            return true;
        }
        if (command.equals(EDIT_COMMAND)) {
            executeEditCommand(attendanceBook);
            return true;
        }
        if (command.equals(CHECK_CREW_ATTEND_COMMAND)) {
            executeCheckAttendCommand(attendanceBook);
            return true;
        }
        if (command.equals(CHECK_WARNING_CREW_COMMAND)) {
            executeCheckWarningCrewsCommand(attendanceBook);
            return true;
        }
        return true;
    }
```
하지만 커맨드별로 메서드를 분리해도 명령어에 따른 분기를 나누고, 이를 처리하는 메서드를 호출하는 형식으로 작성하니 메서드의 길이 10을 초과해 버립니다.
10줄 이내로 줄이는 방법이 떠오르지 않는데 이를 해결하기 위한 설계나 방법이 있을까요?

summerlunaa replied:
연관된 상수가 여러개면 enum으로 관리하는 것도 좋을 것 같아요.

if 문 없이 command에 따라 특정 메서드를 실행하려면 함수형 인터페이스를 사용해 볼 수 있습니다. Map에 key로 command를, value로 함수형 인터페이스를 넣고 command에 따라 key로 함수를 찾아 실행하는 방식으로 수정해볼 수 있겠네요 :) 여기서 더 나아가면 Map이 아니라 enum에 command와 함수를 정의해두고 사용할 수도 있겠구요. 
구체적인 구현 방식은 고민해보세요. 일부러 키워드만 드렸습니다. 시간나면 적용해보세용 😃

summerlunaa commented on src/main/java/domain/AttendanceBook.java:
```diff
@@ -0,0 +1,117 @@
+        for (final String name : map.keySet()) {
+            AttendCount attendCount = checkAttendance(name, days).countAttendStatus();
+            WarningStatus warningStatus = attendCount.judgeWarning();
+            // TODO: Indent 줄이기
```
제출할 때는 TODO는 웬만하면 남기지 않는 것이 좋습니다! 이번에 주석에 적힌대로 리팩토링하면서 주석도 제거해보죠!

eoehd1ek replied:
indent 깊이 문제를 해결하기 위해 작은 단위의 문제로 분리했습니다.
이후 분리한 문제를 해결하기 위한 메서드를 만들어 stream을 사용하도록 해결했습니다.
[refector: 경고 대상자를 가져오는 메서드의 intent를 줄이도록 수정](https://github.com/woowacourse/java-attendance/pull/56/commits/a21c7e28e444cbda08950162d1eed5e0bc184f8c)

좋은 제안 감사합니다.

summerlunaa commented on src/test/java/AttendCountTest.java:
```diff
@@ -0,0 +1,86 @@
+
+    @Test
+    @DisplayName("제적 위험자를 계산하는 기능 - 정상: 무결")
+    void test10() throws Exception {
```
모든 메서드에 불필요한 `throws Exception`은 왜 있는 건가요?
DispayName은 꼼꼼히 잘 적어주셨는데 메서드명은 대충 지은 이유가 있을까요? 전체적으로 DisplayName을 사용한 곳도 있고 아닌 곳도 있어서 통일성 있게 작성해주시면 더 좋을 것 같아요!

eoehd1ek replied:
```java
@Test
void test() throws Exception {}
```
매크로를 사용하여 위와 같은 양식의 코드를 만들어 테스트를 작성했습니다.
초반에 `throws Exception`을 제거하지 않고 커밋을 하여, 나중에 리팩토링 과정에서 일괄적으로 제거하려 했습니다.

DisplayName과 메서드명 부분에선 네이밍을 짓는 시간보다 테스트를 만들고 빠른 구현에 집중할 것인지,
네이밍을 확실히 짓고 테스트 및 구현을 할지 의견이 나뉘었습니다.
그래서 각자 드라이버 차례가 된다면 본인이 원하는 스타일의 테스트 네이밍을 작성했고,
추후 구현 완료 후 DisplayName과 메서드명을 구체화 하기로 했습니다.
그러나 구현 완료 후 리팩터링 할 시간이 없어서 위 상태로 코드 리뷰를 받았습니다.

저의 테스트명 작성 방식은
DisplayName에 한글로 (-값이 주어질 때) (-를 하면, -시) (-한다, -반환한다, -예외가 발생한다) 과 같은 형식으로 작성하려 하며,
이를 참고하여 given, when, then을 작성하려고 합니다. (아니면 given, when, then을 작성 후 이름을 정합니다.)
한글 이름을 정한 후 메서드 명을 작성하며 sholud_동사, should_return_값, should_throw_exception_when_-, 이후 -_by_주어지는_값 과 같은 접미를 붙여 메서드명을 작성합니다.

저는 테스트 메서드명을 짓는데 경험이 적어서 오랜 시간이 걸립니다.
리뷰어님의 테스트 메서드를 작성하는데 이름, DisplayName을 작성하는 과정은 어떻게 되는지 알고싶습니다.
자세하게 코드 리뷰 해 주셔서 감사합니다.

summerlunaa replied:
일단 저는 메서드명을 displayName과 다르게 짓지 않습니다. 영어 작문이 어렵기도 하고 말씀하신대로 한글로도 테스트 네이밍을 짓기 어려운데 영어로 한 번 더 지으려니까 시간도 많이 걸리더라구요. 테스트 메서드명은 displayName과 동일하게 한글로 적으셔도 괜찮을 것 같아요. (메서드명에서는 띄어쓰기가 불가능하니 언더바를 사용합니다)
테스트명은 이미 지금도 잘 작성하고 계신 것 같아요. 저도 given/when/then을 최대한 축약해서 DisplayName으로 작성하는 것이 좋다고 생각합니다 :) 저라고 DisplayName을 작성하는 획기적인 방법을 갖고 있는 건 아니라 답변이 도움이 될지 모르겠네요 😅

summerlunaa commented on src/test/java/AttendsTest.java:
```diff
@@ -0,0 +1,170 @@
+                Attend.of("13", "13:00"));
+        Attends attends = new Attends(new ArrayList<>(attendsInitValue));
+
+        attends.addAttend(Attend.of("16", "10:00"));
+        attends.addAttend(Attend.of("17", "10:17"));
```
테스트는 경계값으로 해주는 게 좋습니다. 현재가 13일이라면 14일로요! 다른 곳들도 경계값으로 테스트하면 더 정확한 테스트가 가능할 것 같네요 👍  전체적으로 테스트 꼼꼼히 잘 작성해주신 것 같습니다. 커밋도 잘 나누셨구요.

eoehd1ek replied:
테스트 시 경계값으로 해야 하는 점 알려주셔서 감사합니다.
첫 리뷰 이후 리뷰 재요청까지 시간이 많이 지나 경계값 테스트로 변경하지 못하고 재요청 드립니다.
다음 리뷰에서 리뷰어님의 테스트 메서드를 작성하는데 이름, DisplayName을 작성하는 과정을 듣고,
참고하여 테스트 메서드명을 짓는 과정을 연습하고 싶습니다.
리뷰를 모두 반영하여 재리뷰 요청 드리지 못한 점 양해 부탁드립니다. 😭

summerlunaa reviewed (changes requested):
후유 안녕하세요~ 파랑입니다 🐳 만나서 반갑습니다!
이번 미션은 요구사항도 많고 TDD도 적용하느라 쉽지 않았을 것 같은데 그래도 잘 구현해주셨습니다!
테스트도 잘 짜셨고 전반적으로 깔끔해서 이해하기 좋았습니다. 컨트롤러를 중심으로 리팩토링 해보시면 좋을 것 같습니다. 이해가 안 되는 부분이 있거나 질문 있으시면 편하게 코멘트나 DM 남겨주세요.
너무 고생하셨습니다 👍

eoehd1ek committed:
refector: 경고 대상자 출력시 중복 사용 코드를 삭제하도록 수정

eoehd1ek committed:
test: 날짜가 일치하는 출석 객체를 찾지 못했을 때 동작하는 누락된 테스트 추가

날짜로 출석 객체를 못 찾으면 예외를 던지는 동작을 변경하며 누락된 테스트를 추가합니다

이후 날짜로 출석 객체를 못 찾으면 Time이 null인 출석 객체를 생성해 반환합니다.

eoehd1ek committed:
refector(Attends): 로직 상 필요 없는 코드를 제거하도록 수정

주어진 Attend에서 날짜만 비교하므로, 날짜로 비교해 동일 출석 가져올 필요가 없다

eoehd1ek committed:
refector: 제적 위험 구하기 로직을 WarningStatus 클래스가 가지도록 수정

eoehd1ek committed:
refector: AttendCount의 WarningStatus 의존을 제거하도록 수정

eoehd1ek committed:
refector: 경고 대상자를 가져오는 메서드의 intent를 줄이도록 수정

eoehd1ek committed:
refector: 커맨드 문자열을 매직 리터럴에서 상수로 변경하도록 수정

eoehd1ek committed:
refector: 무한 반복 로직을 메서드로 분리하도록 수정

eoehd1ek committed:
refector: 무한 반복 로직과 커맨드 입력&처리를 분리하도록 수정

eoehd1ek committed:
refector: 커맨드 입력과 처리를 분리하도록 수정

eoehd1ek committed:
refector: 커맨드 처리를 메서드로 분리하도록 수정

eoehd1ek committed:
refector: 의미를 정확하게 전달하도록 메서드, 변수 이름 수정

eoehd1ek commented:
안녕하세요 파랑! 💙 
먼저 양질의 리뷰 정말 감사드립니다!
다만 첫 리뷰 이후 시간이 많이 지나기도 했고, 스스로 고민 후 해결되지 못한 부분이 있어 모든 리뷰를 반영하지 못하고 재리뷰요청 드립니다. 🥲 

리뷰어님의 테스트 코드 메서드의 이름, DisplayName을 작성하는 과정을 알고싶어요.
그리고 한 번 따라해 보면서 저만의 테스트 코드 작성 규칙을 좀 더 명확히 하고싶습니다.
그래서 이번 리뷰에는 테스트 코드 부분은 살짝 넘어가주셨으면 좋겠어요...ㅎㅎ

그 외 궁금한 점 코멘트로 남겨 놓았습니다.
https://github.com/woowacourse/java-attendance/pull/56#discussion_r1967350651

다시 한 번 좋은 리뷰를 해 주셔서 감사드립니다.

summerlunaa commented on src/main/java/domain/AttendanceBook.java:
```diff
@@ -0,0 +1,123 @@
+
+    private boolean isWarningCrew(String name, List<Integer> days) {
+        WarningStatus warningStatus = judgeWarningStatus(name, days);
+        if (warningStatus == WarningStatus.CLEAR) {
+            return false;
+        }
+        return true;
```
```suggestion
        return warningStatus != WarningStatus.CLEAR;
```

이렇게도 가능하겠네용.

eoehd1ek replied:
파랑께서 알려주신 상수에 final을 붙이지 않아 발생한 경고와 if문 단순화에 대한 경고를 보고 IDE에서 표시되는 경고를 살펴봤습니다.
그러다보니 수정 제안 사항도 함께 표시 된다는 걸 알았습니다.
앞으로는 커밋 전 경고 사항을 확인 후 빠르게 해결 할 수 있는 부분이라면 수정 후 커밋 하도록 하겠습니다.
IDE 기능을 더욱 활용할 수 있도록 제안 해 주셔서 감사합니다.

summerlunaa commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,139 @@
+    }
+
+    private void executeAttendCommand(AttendanceBook attendanceBook) {
+        String nickName = inputView.inputNickName();
+        String time = inputView.inputTime();
+        Attend attend = Attend.of(time);
+        attendanceBook.attend(nickName, attend);
+        AttendStatus attendStatus = attendanceBook.checkAttendance(attend);
+        outputView.printAttendResult(attend, attendStatus);
```
적절한 줄바꿈은 가독성에 도움이 됩니다.

eoehd1ek replied:
제가 봐도 너무 긴 본문에 줄바꿈 없이 코드가 작성되어 있으니 한 눈에 알아보기 쉽지 않네요.
메서드로 분리하거나 줄바꿈 추가를 해야겠습니다.
제안 감사합니다. ❤️

summerlunaa commented on src/main/java/domain/AttendanceBook.java:
```diff
@@ -0,0 +1,123 @@
+import java.util.stream.Collectors;
+import util.DateUtil;
+
+public class AttendanceBook {
+    private static final LocalTime START_TIME = LocalTime.of(8, 0);
+    private static final LocalTime LATE_TIME = LocalTime.of(10, 5);
+    private static final LocalTime ABSENCE_TIME = LocalTime.of(10, 30);
+    private static final LocalTime END_TIME = LocalTime.of(23, 0);
```
지금보니 `교육 시간은 월요일은 13:00~18:00`라는 요구사항이 구현되지 않은 것 같은데 맞나요?

eoehd1ek replied:
앗!...... 그렇네요... 다시 생각해봐도 구현한 적이 없고, 코드를 봐도 해당 요구사항이 구현되지 않았습니다... 😭 
이를 위해 지각, 결석 시간을 가져오는 enum을 구현했습니다.
페어와 코드 작성 중 출석부인 AttendanceBook이 지각, 결석 시간까지 책임져야 하나 고민이 많았는데 분리하니 후련하네요!
코드 하나하나 세세한 리뷰 정말 감사드립니다. 🫶

summerlunaa reviewed (changes requested):
후유 안녕하세요~
리뷰 잘 반영해주셨습니다 👍 변경된 부분과 남겨주신 질문에 대해서 추가로 리뷰 남겨드렸어요.
조금만 더 힘내보시죠ㅎㅎㅎ 궁금한 점 있으면 편하게 DM 주셔도 괜찮습니다. 파이팅!

eoehd1ek committed:
refector: 상수 및 불변하는 인스턴스 변수에 final 키워드 추가

eoehd1ek committed:
refector: if 구문 단순화 수정

eoehd1ek committed:
refector: 람다 표현식으로 수정

eoehd1ek committed:
refector: 필요한 줄바꿈 추가 및 의미있는 변수명으로 수정

eoehd1ek committed:
fix: 출력 양식(줄바꿈) 수정

eoehd1ek committed:
refector: 일급 컬렉션 인스턴스 변수 final 키워드 추가 및 변수명 수정

eoehd1ek committed:
feat: 월요일과 화-금요일의 출결 시작 시간을 다르게 계산하는 기능 추가

eoehd1ek committed:
refector(View): 코드 줄바꿈 양식에 맞게 수정

eoehd1ek commented:
안녕하세요 파랑!
파랑의 테스트 코드 작성법 많은 도움 됐습니다.
다양한 사람의 의견을 들어보고 싶었어요. 🦻 

현재 `교육 시간은 월요일은 13:00~18:00` 부분을 수정했습니다.
그 외 몇몇 간략한 코드 리팩터링 후 다시 리뷰 보냅니다. (항상 늦게, 그리고 늦은 시간에 리뷰 요청을 보내서 미안해요.. 🙏, 통학에 2시간씩 소요되다 보니 집에 가면 기진맥진 해져서 캠퍼스에서만 코드를 작성하게 되네요.. 반성합니다...)

내일까지 enum과 함수형 인터페이스를 사용해 컨트롤러 메서드의 본문을 줄이고,
가능하다면 테스트 코드의 리팩터링까지 완료 후 다시 리뷰를 또 보내겠습니다.
또 즐거운 생각할 거리가 담긴 리뷰 기다리겠습니다! (응원의 말도 환영합니다.)

summerlunaa reviewed (approved):
후유 안녕하세요~
이번에도 리뷰 꼼꼼하게 확인해주시고 반영까지 잘 해주셨군요 👍 추가로 리팩토링하고 싶은 부분들은 2단계에서 반영해보시죠~
1단계는 여기서 마무리 하겠습니다. 수고하셨습니다!
